### DISCUSSION TITLE
Couldn't ingest precipitation from grib2 model file.

### DISCUSSION THREAD

**Discussion by AnastasiaBundel** [dtcenter/METplus#1138] [File I/O] [answered]
Hello,

I try to run GridStat using 1h precipitation accumulated from the initial time from a grib2 model file with the following parameters:

```
grib_ls -p name,max,min,parameter,centre -w shortName=tp ilfr_00010000s.cfoeps.grb
ilfr_00010000s.cfoeps.grb
name                       max         min         parameter   centre
Total Precipitation  2.54492     0            228228       ecmf
1 of 23 grib messages in ilfr_00010000s.cfoeps.grb
```

MET doesn't find any precipitation records (while other records are read correctly, temperature, for example):

```
abundel@xfront7:~/MET/EPS> ./test_grid_stat_radar.sh

*** Running Grid-Stat on precip using GRIB forecast observation ***
DEBUG 1: Default Config File: /RHM-Lustre3.2/users/cosmo/akirsanov/MET/share/met/config/GridStatConfig_default
DEBUG 1: User Config File: /RHM-Lustre3.2/users/cosmo/abundel/MET/EPS/GridStatConfig_radar
GSL_RNG_TYPE=mt19937
GSL_RNG_SEED=1
DEBUG 1: Forecast File: /RHM-Lustre3.2/users/cosmo/abundel/MET/EPS/model/2021070100.CFO2/00/ilfr_00010000s.cfoeps.grb
DEBUG 1: Observation File: /RHM-Lustre3.2/users/cosmo/abundel/MET/EPS/model/2021070100.CFO2/00/ilfr_00020000s.cfoeps.grb
WARNING:
**WARNING: MetGrib2DataFile::data_plane() -> No matching record found for 'APCP/A01'**
WARNING:
WARNING:
WARNING: read_data_plane() -> APCP/A01 not found in file: /RHM-Lustre3.2/users/cosmo/abundel/MET/EPS/model/2021070100.CFO2/00/ilfr_00010000s.cfoeps.grb
WARNING:
ERROR  :
ERROR  : process_scores() -> no reques_ted data found!  Exiting...
ERROR  :

```
What could be the problem? I saw that the precipitation record from a MET example file has the similar parameters:

```
gbf006.gfs.2017061200_tp_only
date                         stepRange                    shortName                    name                         parameterCategory            parameterNumber              parameter                    min                          max
20170612                     0-6                          tp                           Total Precipitation          not_found                    not_found                    228228                       0                            139
1 of 1 messages in pgbf006.gfs.2017061200_tp_only
```
I attach my files in zip archive here.
Thank you,
Anastasia.
[files.ZIP](https://github.com/dtcenter/METplus/files/7106358/files.ZIP)

**Comment by JohnHalleyGotway** [comment, marked as answer]
Hi Anastasia, thanks for sending along some sample data to demonstrate the issue. That made it really easy to debug.

I see that you'd like to run Grid-Stat to evaluate precipitation. When working with GRIB2 files, I use the "wgrib2" utility to dump an inventory of their contents. For example, here's what I see in one of the data files you sent:
```
wgrib2 ilfr_00010000s.cfoeps.grb 
1:0:d=2021070100:PRMSL:mean sea level:60 min fcst:ENS=? table4.6=192 pert=0
2:384520:d=2021070100:PRES:surface:60 min fcst:ENS=? table4.6=192 pert=0
3:769040:d=2021070100:UGRD:10 m above ground:60 min fcst:ENS=? table4.6=192 pert=0
4:1153560:d=2021070100:VGRD:10 m above ground:60 min fcst:ENS=? table4.6=192 pert=0
5:1538080:d=2021070100:GUST:10 m above ground:0-1 hour max fcst:ENS=? table4.6=192 pert=0
6:1922624:d=2021070100:TMP:2 m above ground:60 min fcst:ENS=? table4.6=192 pert=0
7:2307144:d=2021070100:TMP:surface:60 min fcst:ENS=? table4.6=192 pert=0
8:2691664:d=2021070100:SOILTMP:surface:60 min fcst:ENS=? table4.6=192 pert=0
9:3076184:d=2021070100:TMP:2 m above ground:0-1 hour max fcst:ENS=? table4.6=192 pert=0
10:3460728:d=2021070100:TMP:2 m above ground:0-1 hour min fcst:ENS=? table4.6=192 pert=0
11:3845272:d=2021070100:DPT:2 m above ground:60 min fcst:ENS=? table4.6=192 pert=0
12:4229792:d=2021070100:RH:2 m above ground:60 min fcst:ENS=? table4.6=192 pert=0
13:4614312:d=2021070100:SPFH:surface:60 min fcst:ENS=? table4.6=192 pert=0
14:4998832:d=2021070100:CDCC:800 mb - surface:60 min fcst:ENS=? table4.6=192 pert=0
15:5383352:d=2021070100:CDCC:400-800 mb:60 min fcst:ENS=? table4.6=192 pert=0
16:5767872:d=2021070100:CDCC:0-400 mb:60 min fcst:ENS=? table4.6=192 pert=0
17:6152392:d=2021070100:TCDC:surface:60 min fcst:ENS=? table4.6=192 pert=0
18:6536912:d=2021070100:TPRATE:surface:0-1 hour acc fcst:ENS=? table4.6=192 pert=0
19:6921456:d=2021070100:LSRRATE:surface:0-1 hour acc fcst:ENS=? table4.6=192 pert=0
20:7306000:d=2021070100:LSSRWE:surface:0-1 hour acc fcst:ENS=? table4.6=192 pert=0
21:7328842:d=2021070100:CRRATE:surface:0-1 hour acc fcst:ENS=? table4.6=192 pert=0
22:7713386:d=2021070100:CSRWE:surface:0-1 hour acc fcst:ENS=? table4.6=192 pert=0
23:7736228:d=2021070100:WIWW:surface:60 min fcst:ENS=? table4.6=192 pert=0
```
Note that there is no accumulated precipitation, named "APCP", present in that file. And that's why you're seeing this warning:
```
WARNING: MetGrib2DataFile::data_plane() -> No matching record found for 'APCP/A01'
```
However, TPRATE, for total precipitation rate, is present:
```
18:6536912:d=2021070100:TPRATE:surface:0-1 hour acc
```
But that naming convention (TPRATE) is not obvious in the output of the "grib_ls" command you ran, and I suspect that's the source of the confusion.

Whenever getting going with new datasets, I always recommend running MET's plot_data_plane tool:
```
plot_data_plane ilfr_00010000s.cfoeps.grb ilfr_00010000s.cfoeps.TPRATE.ps 'name="TPRATE"; level="A1"; file_type=GRIB2;'
```
Which creates this plot:
<img width="661" alt="Screen Shot 2021-09-03 at 12 32 02 PM" src="https://user-images.githubusercontent.com/21087144/132051359-74bcbd33-bb61-47f1-bb4f-76e8f0a4dbd5.png">

Note that I had to specify "file_type=GRIB2;" since the input file uses the ".grb" suffix which MET interprets as being a GRIB version 1 file. If you renamed it as ".grb2" then you would no longer need that "file_type" option. MET interprets the ".grb2" suffix as meaning its a GRIB2 file.

So you'll need to change "APCP" to "TPRATE" in your Grid-Stat config file. However, if you run Grid-Stat to compare ilfr_00010000s.cfoeps.grb to ilfr_00020000s.cfoeps.grb you'll get a different error message. Because while the first contains the total precip rate accumulated over 1 hour (A1), the second contains it accumulated over 2 hours (A2). I'm guessing that these are just two output timesteps from some model run. So both are forecasts, and I doubt you actually want to compare them directly with Grid-Stat. If you did want to do that for some reason, you'd need to specify the "fcst" and "obs" dictionaries separately in the config file. One would be TPRATE/A1 and the other would be TPRATE/A2.

I'd recommend looking closely at your verifying observation/analysis data. Think about what accumulation interval is present there and adjust your forecast so that your can make a fair comparison. You may find MET's pcp_combine tool useful here. For example, if you want the accumulation between hours 1 and 2, you could run a subtract command like this:

```
pcp_combine -subtract \
ilfr_00020000s.cfoeps.grb 'name="TPRATE"; level="A2"; file_type=GRIB2;' \
ilfr_00010000s.cfoeps.grb 'name="TPRATE"; level="A1"; file_type=GRIB2;' \
ilfr_00010000s.cfoeps.TPRATE_1_2.nc
```

Also note that you can use the "convert(x)" function to convert your hourly rate to a total accumulation. Running plot_data_plane on your 2-hour accumulation is a good example of this:
```
plot_data_plane ilfr_00020000s.cfoeps.grb ilfr_00020000s.cfoeps.APCP.ps \
> 'name="TPRATE"; level="A2"; file_type=GRIB2; convert(x)=x*2;' \
> -title "2-hour Precipitation Accumulation"
```
<img width="594" alt="Screen Shot 2021-09-03 at 12 45 36 PM" src="https://user-images.githubusercontent.com/21087144/132052759-a26dad34-81c9-4ed0-97b8-de2d674951f4.png">
So that convert(x) command transforms the hourly rate to a total accumulation.

Hope that helps clarify.

**Reply by AnastasiaBundel** [reply]
Dear John, 
many thanks for clarification about TPRATE! Great that it works. 
The second model field instead of obs was used as a dummy to debug reading the forecast. Not a good idea to use different accumulations for that though.
Now I'll use the radar data.
Thank you for a pcp_combine example and other tips. I was going certainly to use pcp_combine tool.
Thanks again, 
Anastasia.
